### PR TITLE
[#731] Make blockId layout configured by client side

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

Make blockId layout configurable by client side.

### Why are the changes needed?

To solve the problem mentioned by #731, this PR introduce the config of `rss.client.blockId.layout` to make it possible to configure on the client side.

### Does this PR introduce _any_ user-facing change?

1. Introducing the config of `rss.client.blockId.layout`

### How was this patch tested?
1. UTs

### CODE REVIEW

**Comment by zuston** [comment]
Could you help review this @jerqi

File: common/src/main/java/org/apache/uniffle/common/BlockIdLayoutConfig.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  public static BlockIdLayoutConfig from() {
+    return new BlockIdLayoutConfig(
+        Constants.PARTITION_ID_MAX_LENGTH,
+        Constants.TASK_ATTEMPT_ID_MAX_LENGTH,
+        Constants.ATOMIC_INT_MAX_LENGTH
+    );
+  }
+
+  public static BlockIdLayoutConfig from(int partitionIdLength, int taskAttemptIdLength, int sequenceIdLength) {
+    if (partitionIdLength + taskAttemptIdLength + sequenceIdLength != BLOCK_ID_LENGTH) {
```
Comment (xianjingfeng) [inline common/src/main/java/org/apache/uniffle/common/BlockIdLayoutConfig.java:96]: How about if one of them is 0?

Comment (xianjingfeng) [reply common/src/main/java/org/apache/uniffle/common/BlockIdLayoutConfig.java:96]: And should we put this judge logic into the construtor of BlockIdLayoutConfig?

File: common/src/main/java/org/apache/uniffle/common/BlockIdLayoutConfig.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  private static BlockIdLayoutConfig from(Map<String, String> blockIdLayoutMap) {
+    String partitionIdRawVal = blockIdLayoutMap.get(PARTITION_ID_LENGTH);
+    if (partitionIdRawVal == null) {
+      throw new IllegalArgumentException(PARTITION_ID_LENGTH + " must be configured.");
+    }
+    String taskAttemptIdRawVal = blockIdLayoutMap.get(TASK_ATTEMPT_ID_LENGTH);
+    if (taskAttemptIdRawVal == null) {
+      throw new IllegalArgumentException(TASK_ATTEMPT_ID_LENGTH + " must be configured.");
```
Comment (xianjingfeng) [inline common/src/main/java/org/apache/uniffle/common/BlockIdLayoutConfig.java:109]: How about use `assert`?

**Comment by jerqi** [comment]
It's dangerous that let users configure the length directly. Maybe we could provide several options for them, such as `BIG_JOB`, `LONG_TIME_JOB`, etc.

**Comment by zuston** [comment]
> It's dangerous that let users configure the length directly. Maybe we could provide several options for them, such as `BIG_JOB`, `LONG_TIME_JOB`, etc.

SGTM.

**Comment by jerqi** [comment]
We need reconsider this pr. Task can cycle and layout can be configured. We should think more.

**Comment by connorlwilkes** [comment]
Any movement on this particular feature?

**Comment by zuston** [comment]
> Any movement on this particular feature?

No, do you want to pick up this?

**Comment by zuston** [comment]
> The reason this is possible is that the block ids are bound to shuffle ids, it shouldn't interfere with each other.

But we also need to get all valid blockIds by the upstream map Ids in client.

**Pull request closed without merging**